src/str.rs

```diff
@@ -0,0 +1,139 @@
+#[derive(Debug)]
+pub enum EscapedStringFragment<'a> {
+    NotEscaped(&'a str),
+    Escaped(char),
+}
+
+#[derive(Debug)]
+pub enum StringUnescapeError {
+    InvalidEscapeSequence,
+}
+
+fn unescape_next_fragment(
+    escaped_string: &str,
+) -> Result<(EscapedStringFragment<'_>, &str), StringUnescapeError> {
+    Ok(if let Some(rest) = escaped_string.strip_prefix('\\') {
+        let mut escaped_string_chars = rest.chars();
+
+        let unescaped_char = match escaped_string_chars.next() {
+            Some('"') => '"',
+            Some('\\') => '\\',
+            Some('/') => '/',
+            Some('b') => '\x08',
+            Some('f') => '\x0C',
+            Some('n') => '\n',
+            Some('r') => '\r',
+            Some('t') => '\t',
+            Some('u') => {
+                fn split_first_slice(s: &str, len: usize) -> Option<(&str, &str)> {
+                    Some((s.get(..len)?, s.get(len..)?))
+                }
+
+                let (escape_sequence, remaining_escaped_string_chars) =
+                    split_first_slice(escaped_string_chars.as_str(), 4)
+                        .ok_or(StringUnescapeError::InvalidEscapeSequence)?;
+
+                escaped_string_chars = remaining_escaped_string_chars.chars();
+
+                u32::from_str_radix(escape_sequence, 16)
+                    .ok()
+                    .and_then(char::from_u32)
+                    .ok_or(StringUnescapeError::InvalidEscapeSequence)?
+            }
+            _ => return Err(StringUnescapeError::InvalidEscapeSequence),
+        };
+
+        (
+            EscapedStringFragment::Escaped(unescaped_char),
+            escaped_string_chars.as_str(),
+        )
+    } else {
+        let (fragment, rest) =
+            escaped_string.split_at(escaped_string.find('\\').unwrap_or(escaped_string.len()));
+
+        (EscapedStringFragment::NotEscaped(fragment), rest)
+    })
+}
+
+pub(crate) fn unescape_fragments(
+    mut escaped_string: &str,
+) -> impl Iterator<Item = Result<EscapedStringFragment<'_>, StringUnescapeError>> {
+    core::iter::from_fn(move || {
+        if escaped_string.is_empty() {
+            None
+        } else {
+            Some(
+                unescape_next_fragment(escaped_string).map(|(fragment, rest)| {
+                    escaped_string = rest;
+                    fragment
+                }),
+            )
+        }
+    })
+}
+
+/// A borrowed escaped string
+#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(rename = "__serde_json_core_escaped_string__")]
+pub struct EscapedStr<'a>(&'a str);
+
+impl<'a> EscapedStr<'a> {
+    pub(crate) const NAME: &'static str = "__serde_json_core_escaped_string__";
+
+    /// Create a new EscapedString, verifying that it's correctly escaped
+    pub fn new(escaped_string: &'a str) -> Result<Self, StringUnescapeError> {
+        // Check that all fragments are valid
+        for fragment in unescape_fragments(escaped_string) {
```

Comment:
Why bother checking in advance if we are going to do a result propogation at the higher level anyways? I'd recommend we remove this since its just processing overhead.